src/util/timemanager.cpp

```diff
@@ -59,133 +121,264 @@ void TimeManager::preSynchronization(double dt) {
-            return;
-        }
+void TimeManager::progressTime(double dt) {
+    static_cast<Time&>(_integrateFromTime) = _currentTime;
+    // Frames     |    1                    2          |
+    //            |------------------------------------|
+    // Keyframes  | a     b             c       d   e  |
+
+    // 1: Previous frame
+    // 2: Current frame (now)
+    // lastPastKeyframe: c
+    // firstFutureKeyframe: d
+    // _latestConsumedTimestamp: a.timestamp
+
+    if (_shouldSetTime) {
+        // Setting the time using `setTimeNextFrame`
+        // will override any timeline operations.
+        static_cast<Time&>(_currentTime).setTime(_timeNextFrame.j2000Seconds());
+        static_cast<Time&>(_integrateFromTime).setTime(_timeNextFrame.j2000Seconds());
+        _shouldSetTime = false;
 
-        if (time().paused()) {
-            return;
+        using K = const CallbackHandle;
+        using V = TimeChangeCallback;
+        for (const std::pair<K, V>& it : _timeJumpCallbacks) {
+            it.second();
         }
+        return;
+    }
+
+    const double now = OsEng.windowWrapper().applicationTime();
+    const std::deque<Keyframe<TimeKeyframeData>>& keyframes = _timeline.keyframes();
 
-        const double secondsOffTolerance = OsEng.parallelPeer().timeTolerance();
+    auto firstFutureKeyframe = std::lower_bound(
+        keyframes.begin(),
+        keyframes.end(),
+        now,
+        &compareKeyframeTimeWithTime
+    );
 
-        const double predictedTime = time().j2000Seconds() + time().deltaTime() *
-                                     (next.timestamp - now);
-        const bool withinTolerance = std::abs(predictedTime - nextTime.j2000Seconds()) <
-                                     std::abs(nextTime.deltaTime() * secondsOffTolerance);
+    const bool hasFutureKeyframes = firstFutureKeyframe != keyframes.end();
+    const bool hasPastKeyframes = firstFutureKeyframe != keyframes.begin();
+
+    const auto lastPastKeyframe = hasPastKeyframes ?
+        (firstFutureKeyframe - 1) :
+        keyframes.end();
+
+    const bool hasConsumedLastPastKeyframe = hasPastKeyframes ?
+        (lastPastKeyframe->timestamp <= _latestConsumedTimestamp) :
+        true;
+
+    if (hasFutureKeyframes && hasPastKeyframes && !firstFutureKeyframe->data.jump) {
+        // If keyframes exist before and after this frame,
+        // interpolate between those.
+        TimeKeyframeData interpolated =
+            interpolate(*lastPastKeyframe, *firstFutureKeyframe, now);
+        static_cast<Time&>(_currentTime).setTime(interpolated.time.j2000Seconds());
+        _deltaTime = interpolated.delta;
+    } else if (!hasConsumedLastPastKeyframe) {
+        applyKeyframeData(lastPastKeyframe->data);
+    } else if (!isPaused()) {
+        // If there are no keyframes to consider
+        // and time is not paused, just advance time.
+        _deltaTime = _targetDeltaTime;
+        static_cast<Time&>(_currentTime).advanceTime(dt * _deltaTime);
```

Comment:
The multiple static_cast<Time&> are confusing.  Is there a better way to get the data out of the `SyncData` class?